waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -728,13 +738,24 @@
         k8s-name (service-id->k8s-app-name scheduler service-id)
         health-check-scheme (-> (or health-check-proto backend-proto) http-utils/backend-proto->scheme string/upper-case)
         health-check-url (sd/service-description->health-check-url service-description)
-        memory (str mem "Mi")]
+        memory (str mem "Mi")
+        service-hash (service-id->service-hash service-id)]
     (cond->
       {:kind "ReplicaSet"
        :apiVersion replicaset-api-version
-       :metadata {:annotations {:waiter/service-id service-id}
+       :metadata {;; Since there are length restrictions on Kubernetes label values,
+                  ;; we store just the 32-char hash portion of the service-id as a searchable label,
+                  ;; but store the full service-id as an annotation.
+                  ;; https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
+                  ;; https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set
+                  :annotations {:waiter/service-id service-id}
                   :labels {:app k8s-name
+                           ;; TODO - remove waiter-cluster
+                           ;; after waiter/cluster is exclusively in use
+                           ;; (see GitHub issue #721)
                            :waiter-cluster cluster-name
+                           :waiter/cluster cluster-name
+                           :waiter/service-hash service-hash
                            :waiter/user run-as-user}
```

Comment:
In a separate PR, we should perform a length check on the `run-as-user` label and trim accordingly.

Reply:
The OS ensures that isn't a problem: user names already have a length limit of 32 chars on posix systems.

https://serverfault.com/questions/294121/what-is-the-maximum-username-length-on-current-gnu-linux-systems